### ISSUE TITLE
Multi-net : one net for opening, another for midgame, another for endgame

### ISSUE THREAD

**Issue by ihavnoid** [gcp/leela-zero#1129]
What we can learn from #1113, IMHO, is that a good net still buys us a lot of strength even with the same data.  However, a good net is expensive in terms of compute power, and hence I was thinking if having multiple nets with different sets of training data will help (or hurt).

...and hence decided to conduct a experiment.  I am going to create three 10x128 nets with different sets of data.  First net will be used if the number of player's stone is 0 ~ 50, second net if 51~100, and for everything else, use the third net.  This will also need modifications on the neural net code that it needs multiple networks, and I'll start writing code while the training is running.

Will keep you guys updated!  Any comments/suggestions welcome.

**Comment by hydrogenpi** [comment]
Can you document your procedure steps by steps

**Comment by ihavnoid** [comment]
I will have the patch with comments on the procedure once ready.

**Comment by hydrogenpi** [comment]
What do you think about the adding a liberty plane idea? There are talks over at some minigo groups that if Leelz added it could make it instantly as strong as FineArt level

**Comment by TableCarpet** [comment]
What about completely separate net for black and for white?
Train black until it wins against white. Then train white until it wins against black. And so on.
Also Test Matches becomes more simple, because new net always plays one color.

**Comment by ihavnoid** [comment]
@TableCarpet yes I also thought about it... the thing is, there is nothing inherent on the rule where black is different from white, except the Komi.  That is, I sorta felt like separate nets for black/white isn't going to buy us a lot of network simplification simply because they are too similar...

Maybe not.  Nothing to prove/disprove maybe I will try that too at some point.

**Comment by ihavnoid** [comment]
For the liberty plane idea, I understand it can be interesting but just feels like a completely separate idea which is out of scope of what I was planning.  This is just to see if there is any value of using multiple nets depending on different circumstances.

**Comment by Eddh** [comment]
I think it might be interesting to use the 20 block net at the root node  only. Time lost would be minimal but the strength gain might not be

**Comment by Atarust** [comment]
Moving away from the simplicity of having one net only can introduce problems.
How will you switch between the nets? A "50 moves" hard boundary between the "fuseki-net" and the "middle-game" net seems wrong, as the fuseki may end earlier or later.

The agent becomes exploitable: If I know the middle game net will start from move 50, I will try to exploit the weaknesses of the fuseki net before move 50.

So even if the multi-net idea would give better experimental results, in the long term introducing specifific nets seems the wrong direction to go.

**Comment by ihavnoid** [comment]
Yeah it will have problems.  That's why we still have to do some training beyond the moves that we are going to actually use.  That is, the net for move 0~100 still needs some training on move 100~200.  The idea is not to learn the late moves at all, it's just that we spend less time learning late moves.

**Comment by wpstmxhs** [comment]
Using multiple networks at once, should be not so difficult. I already did it before( #945 ).
This took ~30 minutes to change it.
You just have to change OpenCL file's global variables to class private member, and change Network and OpenCL class to use them.
For your reference, I upload my code.

[LZNew - 6 and 20.zip](https://github.com/gcp/leela-zero/files/1865614/LZNew.-.6.and.20.zip)

Take a look into my old(based on leela zero version which is a few months ago) code.
It uses 6 block and 20 block networks at once, and it mixes two policy head outputs by averaging with codes, and uses 20 block net's value head.

```c++
bool UCTNode::create_children(std::atomic<int> & nodecount,
                              GameState & state,
                              float & eval) {
// ...

    const auto raw_netlist_6b = net_6b.get_scored_moves(
		&state, Network::Ensemble::DIRECT, 0);
    const auto raw_netlist_20b = net_20b.get_scored_moves(
        &state, Network::Ensemble::DIRECT, 0);

    // DCNN returns winrate as side to move
    auto net_eval = raw_netlist_20b.second;

// ...

    std::vector<Network::scored_node> nodelist;

    auto legal_sum = 0.0f;
	const auto size = raw_netlist_20b.first.size();
	for (int i = 0; i < size; ++i) {
		Network::scored_node node = raw_netlist_20b.first[i];
		// mix 6b
		node.first = (node.first + raw_netlist_6b.first[i].first) / 2;
        auto vertex = node.second;
		auto xy = state.board.get_xy(vertex);
        if (state.is_move_legal(to_move, vertex) && !ladder[POS(xy.first + BOARD_START, xy.second + BOARD_START)]) {
            nodelist.emplace_back(node);
            legal_sum += node.first;
        }
    }

// ...

}
```

**Comment by wpstmxhs** [comment]
@gcp stated that he will not accept my supposed change unless it has some advantages.

**Comment by wpstmxhs** [comment]
@Eddh To extend your idea, it should be fun to try to use larger network for many-visited nodes only.

**Comment by lightvector** [comment]
@ihavnoid or anyone else interested - I've posted elsewhere about this before, but it seems very relevant here:

It's very easy to give the neural net the capacity __to learn to detect and adapt to the global game phase all by itself__ by adding a simple structure into the neural net. This should give you much better regularization than having to train 3 separate neural nets, and should be much simpler because it requires no changes at all the training and self-play pipeline.

Into the middle of the neural net (say, halfway deep in the stack of residual blocks), you simply add a single residual block of the following form:

    def global_pooling_residual_block(trunk, num_channels, num_pooled_channels):
        #Compute some channels to be globally mean-pooled so that in later layers the neural net can make its local policy depend on the global status of the board
        layer_to_be_pooled = relu(batchnorm(conv3x3(trunk, num_pooled_channels)))  #shape = batch_size,19,19,num_pooled_channels
        pooled = reduce_mean(layer_to_be_pooled, axis=(1,2), keepdims=True) #shape = batch_size,1,1,pooled_channels
        #Apply linear transform to convert from pooling feature space with num_pooled_channels to normal feature space with num_channels.
        biases= matrix_multiply(pooled, weight_matrix_parameter(pooled_channels, num_channels))  #shape = batch_size,1,1,num_channels

        #Now the rest is just a normal residual block except that we add in the pooled result as biases for the first convolution. 
        #The addition will broadcast from  shape = batch_size,1,1,num_channels -> shape = batch_size,19,19,num_channels
        intermediate_layer = relu(batchnorm(biases+conv3x3(trunk, num_channels))) #shape = batch_size,19,19,num_channels
        
        residual = relu(batchnorm(conv3x3(intermediate_layer, num_channels)))
        return trunk + residual

So now, if there are any global properties that the neural net finds relevant to evaluating a local situation, such as phase of the game (opening, midgame, endgame), the neural net will have the ability to learn them all by itself. _And no specific human knowledge about Go is added, the neural net is just given the power on its own to choose anything it finds important._

In my experiments, if you add such a structure, the neural net does indeed use it to detect game phase and adjust its play accordingly. When visualizing the heatmaps of the channels just-prior to pooling it is completely obvious that some of the channels are game-phase detectors, such as channels that rapidly become more active and saturate as the board fills up (when mean-pooled, detects whether it is the early opening or not), or channels that are only active around unstable areas and weak groups (when mean-pooled, detects whether it is the midgame or the endgame). The neural net also typically uses one or more of the channels to detect if there is a ko ongoing (when mean-pooled, lets the neural net globally know when to look for ko threats).

**Comment by ihavnoid** [comment]
Current progress:

- Since there is no easy way to parse the number of moves played from our training data, I am using the number of the current player's stone as a proxy of move number.
- Using the number of stones, I am selecting samples with a random probability of 1 - (num_stones*num_stones/10000) which means it should be mostly board states with 0~70 stones from the current player.  Translating to mostly early game with decent amount of midgame.
- Dataset is from #167  0fb68ccf ~ 193437be - roughly 80GB worth of data.

With a batch size of 64, learning rate of 0.001 for the first 1.2M steps and 0.0001 for the next 1.2M steps, this is the current state:

```
step 2400000, policy=1.67477 training accuracy=60.1406%, mse=0.22938
```

Accuracy seems to be good but mse seems to be much worse than full training.  I think it's because it consists of more early game states (which should be trained to somewhere around 50%.)  I was hoping for something better, though.

I'm gonna let this run until 2.5~3M steps and then move to lategame, and then train a reference 'full data' net.  Still have to think how to generate the midgame dataset.

**Comment by pheasant75** [comment]
I proposed the idea of training a network that is specialized in end-game in some earlier thread.  Basically we first randomly select any game that is close. Start from play 150 (hopefully it is in endgame stage).  Add noise to the winning side to see if the margin of victory can be increased.  Add noise to losing side to see if it can reduce margin of loss or even reverse it.  Train new network to learn from that.

Or the problem maybe much simpler.  We just add more playouts when we do those self-plays to get higher quality games to learn from.  

I believe the end game should be relatively easier and especially plays into the strength of computer.  By limiting the scenario to end-game, we should be able to generate a strong network specialized in end-game much sooner.

**Comment by lightvector** [comment]
Just to be clear - I'm probably not going to be available myself for testing the idea I suggested above for Leela Zero any time soon, since I'm not set up to do much with Leela Zero and my gpus are already saturated with a project of my own. Kudos to ihavnoid and the others who actually are putting in the time and effort to actually test out potential improvements, which is really where a lot of the work and value is. 

If anyone at any point wants to also try out this sort of neural net structure for Leela Zero and open an issue to report the results, feel free to go right ahead. It would be very interesting if it didn't lead to at least a noticeable gain, given the improvement I observed in my own project from trying it.

**Comment by ihavnoid** [comment]
Progress update:

Created three nets, early / endgame / full data.  No midgame net yet, but still these should be enough to get started.

```
90/90 games played

                     A    B     C
A leelaz-single-net      9-21  9-21
B leelaz-two-net    21-9       20-10
C leelaz-three-net  21-9 10-20
```

single net : only use the full dataset net
two net : use the early game and endgame net only based on number of current player's stones (threshold is 75)
three net : use the early game, full net, and endgame net with thresholds of 50 and 100

Next step is to actually create a midgame net and put it on the three net config.

**Comment by ihavnoid** [comment]
Created a 'midgame' net.  Now the 'three net' actually is an early/mid/endgame net.

4000 playouts
```
                     A    B     C
A leelaz-single-net      8-32  5-35
B leelaz-two-net    32-8       16-24
C leelaz-three-net  35-5 24-16

leelaz-single-net v leelaz-two-net (40/40 games)
board size: 19   komi: 7.5
                    wins              black         white       avg cpu
leelaz-single-net      8 20.00%       4  20.00%     4  20.00%    182.46
leelaz-two-net        32 80.00%       16 80.00%     16 80.00%    169.55
                                      20 50.00%     20 50.00%

leelaz-single-net v leelaz-three-net (40/40 games)
board size: 19   komi: 7.5
                    wins              black         white       avg cpu
leelaz-single-net      5 12.50%       3  15.00%     2  10.00%    181.73
leelaz-three-net      35 87.50%       18 90.00%     17 85.00%    168.47
                                      21 52.50%     19 47.50%

leelaz-two-net v leelaz-three-net (40/40 games)
board size: 19   komi: 7.5
                   wins              black         white       avg cpu
leelaz-two-net       16 40.00%       9  45.00%     7  35.00%    185.03
leelaz-three-net     24 60.00%       13 65.00%     11 55.00%    182.38
                                     22 55.00%     18 45.00%

```

8000 playouts
```
                     A    B     C
A leelaz-single-net      9-31  2-38
B leelaz-two-net    31-9       12-28
C leelaz-three-net  38-2 28-12

leelaz-single-net v leelaz-two-net (40/40 games)
board size: 19   komi: 7.5
                    wins              black         white       avg cpu
leelaz-single-net      9 22.50%       1   5.00%     8  40.00%    634.09
leelaz-two-net        31 77.50%       12 60.00%     19 95.00%    571.56
                                      13 32.50%     27 67.50%

leelaz-single-net v leelaz-three-net (40/40 games)
board size: 19   komi: 7.5
                    wins              black         white       avg cpu
leelaz-single-net      2  5.00%       1   5.00%     1   5.00%    599.74
leelaz-three-net      38 95.00%       19 95.00%     19 95.00%    529.53
                                      20 50.00%     20 50.00%

leelaz-two-net v leelaz-three-net (40/40 games)
board size: 19   komi: 7.5
                   wins              black         white       avg cpu
leelaz-two-net       12 30.00%       4  20.00%     8  40.00%    666.80
leelaz-three-net     28 70.00%       12 60.00%     16 80.00%    645.70
                                     16 40.00%     24 60.00%
```
Seems like an effective strategy.  Of course the hard part is to actually maintain/train 3 different nets.

**Comment by tapsika** [comment]
> Seems like an effective strategy. Of course the hard part is to actually maintain/train 3 different nets.

I think you need to compare to a bigger net (than those 3 individually), since that's what it effectively is.

**Comment by Pytq** [comment]
> I think you need to compare to a bigger net (than those 3 individually), since that's what it effectively is.

Why do you think that ? I think the bottleneck is the selfplay game generation not the training so it makes sense that his comparaison is for equal selfplay speed (and not equal training speed).

**Comment by isty2e** [comment]
I think if this approach works, it would only mean that the training data themselves are okay, but we are close to the capacity of the 128x10 network size.

**Comment by tapsika** [comment]
@Pytq I meant network cognitive capacity. Using different nets for different phases means more effective filter count. This may be viable as a speed-optimization strategy (always trying to select the most useful subnet), but it still seems better to test on both speed parity and network size parity, for a more unbiased picture.

**Comment by Yakago** [comment]
It certainly looks interesting :) - If only as a way to produce a strong network once we're at some point of end point wrt. the regular training

You could perhaps test it against a network from a point that corresponds to your data?

**Comment by ainu7** [comment]
Looks impressive without sacrificing speed!

Is smooth transitioning also possible? i.e, using weighted average of two networks.

**Comment by holicst** [comment]
Wow, congrats! As a next step (if somehow we can have more resources) it'd be great to optimize the number of nets (maybe between 2-5?) and the move-count intervals (that define which net to use). Hopefully if you'll have the data and evidence this can get into the main stream of LZ.

**Comment by isty2e** [comment]
Why not train another network to determine which phase it is? :)

**Comment by ihavnoid** [comment]
I never did think this would have been a good idea to be merged into mainstream, for a couple of reasons:
- This diverges quite a bit from the AlphaGo Zero paper
- Even if we think it is a good idea, it adds pressure to the training pipeline because now we have to train three nets
- With no net, having the code to support multiple nets turns into a bit odd feature.

I was thinking, it might make sense we revive this idea once we sort-of declare that we are done with further training, because we aren't going to move to bigger nets due to becoming too big to run on consumer-level hardware.  We still have a long way to get there, though. :)

**Comment by emc2314** [comment]
I myself cannot agree with the idea to use multi-net for one game. There are several reasons:

- These terms, "early/mid/endgame", are actually discovered by us human. Since we are able to conclude different strategies for different phases, the machine is able to do that too. What we should do is helping machine realize it instead of just training 3 networks.
- Different networks will think differently. If you use hard boundary, MCTS done by different networks near that boundary will vary sharply. Seems like Trump takes over Obama or a dynasty falls.
- Hard to train. A good opening may fail because of bad endgame network. And for the idea to train another network to determine which phase it is, I can't figure out the detailed algorithm. Also the training data set must be much larger compared to current method.
- Unnatural. Zero means zero human knowledge. This idea is not that beautiful.
